hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/HiveSchemaWithPartnerVisitor.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hive.legacy;
+
+import java.util.List;
+import org.apache.hadoop.hive.serde2.typeinfo.ListTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.MapTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.PrimitiveTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.StructTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/**
+ * A Hive {@link TypeInfo} visitor with an accompanying partner schema
+ *
+ * This visitor traverses the Hive {@link TypeInfo} tree contiguously accessing the schema tree for the partner schema
+ * using {@link PartnerAccessors}. When visiting each type in the Hive tree, the implementation is also presented
+ * with the corresponding type from the partner schema, or else a {@code null} if no match was found. Matching
+ * behavior can be controlled by implementing the methods in {@link PartnerAccessors}
+ *
+ * @param <P> type of partner schema
+ * @param <FP> type of the field representation in the partner schema
+ * @param <R> type of the resultant schema generated by the visitor
+ * @param <FR> type of the field representation in the resultant schema
+ */
+@SuppressWarnings("ClassTypeParameterName")
+public abstract class HiveSchemaWithPartnerVisitor<P, FP, R, FR> {
+
+  /**
+   * Methods to access types in the partner schema corresponding to types in the Hive schema being traversed
+   *
+   * @param <P> type of partner schema
+   * @param <FP> type of the field representation in the partner schema
+   */
+  public interface PartnerAccessors<P, FP> {
+
+    FP fieldPartner(P partnerStruct, String fieldName);
+
+    P fieldType(FP partnerField);
+
+    P mapKeyPartner(P partnerMap);
+
+    P mapValuePartner(P partnerMap);
+
+    P listElementPartner(P partnerList);
+  }
```

Comment:
I would remove the name `partner` from the method names and variable names (and in the implementation too) since it is implied from the class name. We might also bring those methods directly to `HiveSchemaWithPartnerVisitor`. We can do this after check-in.

Reply:
Lets followup on this after the check-in. I may not have the same opinion and don't want to delay the rollout since this seems like a minor issue.